operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx context.Context
+
+	// Only used during testing.
+	initDoneCallbacks []LBIPAMInitDone
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*core_v1.Service]
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	ipam.ctx = ctx
+
+	errChan := make(chan error, 2)
+
+	poolChan := stream.ToChannel[ipPoolEvent](ctx, errChan, ipam.poolResource)
+	svcChan := stream.ToChannel[svcEvent](ctx, errChan, ipam.svcResource)
+
+	err := ipam.init(ctx, poolChan, svcChan)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while initializing LB-IPAM")
+		// Stop, don't go into main control loop
+		return
+	}
+
+	for _, cb := range ipam.initDoneCallbacks {
+		if cb != nil {
+			cb()
+		}
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are closed to avoid blocking
+	// the resources in case they are re-used after we shutdown.
+	for poolChan != nil && svcChan != nil {
+		select {
+		case event, ok := <-poolChan:
+			if !ok {
+				poolChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.poolOnUpsert,
+				ipam.poolOnDelete,
+			)
+		case event, ok := <-svcChan:
+			if !ok {
+				svcChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.svcOnUpsert,
+				ipam.svcOnDelete,
+			)
+		case err := <-errChan:
+			ipam.logger.WithError(err).Error("Resource error")
+		}
+	}
+}
+
+func (ipam *LBIPAM) init(
+	ctx context.Context,
+	poolChan <-chan ipPoolEvent,
+	svcChan <-chan svcEvent,
+) error {
+	var syncErrors []error
+
+	// First sync all pools
+	poolSync := false
+	for !poolSync {
+		select {
+		case event := <-poolChan:
+			event.Handle(
+				func() error {
+					err := ipam.settleConflicts()
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+					}
+
+					poolSync = true
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					err := ipam.handleNewPool(clbi)
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+					}
+
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					syncErrors = append(syncErrors, errors.New("Got a Pool deleted event while syncing"))
+
+					return nil
+				},
+			)
+		}
+	}
+
+	// Then sync all services
+	svcSync := false
+	for !svcSync {
+		select {
+		case event := <-svcChan:
+			event.Handle(
+				func() error {
+					svcSync = true
+
+					err := ipam.satisfyServices()
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+						return nil
+					}
+
+					err = ipam.updateAllPoolCounts()
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+						return nil
+					}
+
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					err := ipam.handleUpsertService(svc)
+					if err != nil {
+						syncErrors = append(syncErrors, err)
```

Comment:
Why wouldn't you want to re-try here? What if there's a brief blip in apiserver connectivity? Surely we should just retry in that case, right?

Reply:
Yep, I'd also just retry all errors instead (e.g. return non-nil from handler). Seems like we just now return the error, log it and then just stop which may cause a transient connectivity failure to leave the operator in a broken state requiring the admin to manually restart the operator process.

Reply:
Agreed. I now implemented custom retry logic for the pools. This is because I need to be 100% sure that all pools are synced without errors before continuing to process services, and the resource retry mechanism doesn't have a way to give that guarantee. For the services I simply retry using the resources built-in retry mechanism.